core/domain/exp_jobs_one_off.py

```diff
@@ -728,3 +728,179 @@ def map(item):
     @staticmethod
     def reduce(key, values):
         yield (key, values)
+
+
+class ExpSnapshotsMigrationAuditJob(jobs.BaseMapReduceOneOffJobManager):
+    """A reusable one-off job for testing the migration of all exp versions
+    to the latest schema version. This job runs the state migration, but does
+    not commit the new exploration to the datastore.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationSnapshotContentModel]
+
+    @classmethod
+    def enqueue(cls, job_id, additional_job_params=None):
+        super(ExpSnapshotsMigrationAuditJob, cls).enqueue(
+            job_id, shard_count=64)
+
+    @staticmethod
+    def map(item):
+        exp_id = item.get_unversioned_instance_id()
+
+        latest_exploration = exp_fetchers.get_exploration_by_id(
+            exp_id, strict=False)
+        if latest_exploration is None:
+            yield ('INFO - Exploration does not exist', item.id)
+            return
+
+        exploration_model = exp_models.ExplorationModel.get(exp_id)
+        if (exploration_model.states_schema_version !=
+                feconf.CURRENT_STATE_SCHEMA_VERSION):
+            yield (
+                'FAILURE - Exploration is not at latest schema version', exp_id)
+            return
+
+        try:
+            latest_exploration.validate()
+        except Exception as e:
+            yield (
+                'INFO - Exploration %s failed non-strict validation' % item.id,
+                e)
+
+        # Some old explorations do not have a states schema version.
+        if 'states_schema_version' not in item.content:
+            item.content['states_schema_version'] = 0
+
+        target_state_schema_version = feconf.CURRENT_STATE_SCHEMA_VERSION
+        current_state_schema_version = item.content['states_schema_version']
+        if current_state_schema_version == target_state_schema_version:
+            yield (
+                'SUCCESS - Snapshot is already at latest schema version',
+                item.id)
+            return
+
+        versioned_exploration_states = {
+            'states_schema_version': current_state_schema_version,
+            'states': item.content['states']
+        }
+        while current_state_schema_version < target_state_schema_version:
+            try:
+                assert (
+                    versioned_exploration_states['states_schema_version'] ==
+                    current_state_schema_version)
```

Comment:
Is this necessary? Seems like an if-else is more appropriate here.

Reply:
It's not entirely necessary. Just a sanity-check to make sure the dict is in sync with the counter, since the info is provided in two places. I can remove it if you like, but an if-else is not really the way to go here since it's important that the "else" case should not occur.

The double-source-of-truth is an issue with the existing migration system and should probably be fixed separately. This PR does not attempt to do that.

Reply:
I'm not sure I understand, is there a test case that demonstrates why this pattern is needed explicitly?

Reply:
To be clear, the only thing I feel strongly about here is using `assert` for control flow. If/for/else can do the job just-as-well, so I'd prefer we restructure the code to avoid using it.

Reply:
Ack. Actually I think I can just drop the assert. Would that be ok? On reflection, I don't think I need it.

I'm actually in agreement with what you're saying in general, too (re the for loop). I'm just a bit reluctant to modify it too much because (a) this follows the pattern of the existing migration audit job in this file, and (b) I've tested this job on the server already (and it's quite expensive). Do you think removing the assert is sufficient or should I make the bigger change?

Reply:
Done (here and in 3 other places). Thanks!